Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,640 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+using MediatR;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.Notifications {
+  /// <summary>
+  /// DTO used to communicate the current verification diagnostics to the LSP client.
+  /// </summary>
+  [Method(DafnyRequestNames.VerificationDiagnostics, Direction.ServerToClient)]
+  public class VerificationDiagnosticsParams : IRequest, IRequest<Unit> {
+    public VerificationDiagnosticsParams(DocumentUri uri,
+      int version,
+      VerificationTree[] verificationTrees,
+      Container<Diagnostic> diagnostics,
+      int linesCount,
+      bool verificationStarted,
+      int numberOfResolutionErrors) {
+      Uri = uri;
+      Version = version;
+      VerificationTrees = verificationTrees;
+      Diagnostics = diagnostics;
+      if (linesCount != 0) { // Deserialization makes linesCount to be equal to zero.
+        PerLineDiagnostic =
+          RenderPerLineDiagnostics(this, verificationTrees, linesCount, numberOfResolutionErrors, verificationStarted, diagnostics);
+      }
+    }
+
+    /// <summary>
+    /// Gets the URI of the document whose verification completed.
+    /// </summary>
+    public DocumentUri Uri { get; init; }
+
+    /// <summary>
+    /// Gets the version of the document.
+    /// </summary>
+    public int? Version { get; init; }
+
+    /// <summary>
+    /// Gets the same diagnostics as displayed in the diagnostics window
+    /// </summary>
+    public Container<Diagnostic> Diagnostics { get; init; }
+
+    /// <summary>
+    /// Returns a tree of diagnostics that can be used
+    /// First-level nodes are methods, or witness subset type checks
+    /// Second-level nodes are preconditions, postconditions, body verification status
+    /// Third level nodes are assertions inside functions 
+    /// </summary>
+    public VerificationTree[] VerificationTrees { get; init; }
```

Comment:
As indicated in the PR description: _"Send verification diagnostics along with per-line status to the client for display"_, this PR is not just scoped for showing icons in the gutter, but to also display dynamic relationships between errors (e.g. return paths and postconditions) and also the verification trace, for which the verification tree is useful because it contains all this information.

I opens the way to displaying the verification tree later during extension development.

Reply:
> to also display dynamic relationships between errors (e.g. return paths and postconditions)

What part of the tree is used for that?

> It opens the way to displaying the verification tree later during extension development.

I'd prefer leaving that out of scope for this PR. We should think about what an efficient way to update the tree is. Maybe we should have a request from the client that sends over the entire tree, and then notifications that only update subtrees within the trees.

Reply:
> What part of the tree is used for that?

relatedRanges (for the verification trace), immediatelyRelatedRanges (for the errors in the same context as the current error range, such as the return path), and dynamicallyRelatedRanges (for ranges that would otherwise be displayed as "Related location" in the diagnostics but happen outside of the context, i.e. for failing call preconditions)

> I'd prefer leaving that out of scope for this PR. We should think about what an efficient way to update the tree is. Maybe we should have a request from the client that sends over the entire tree, and then notifications that only update subtrees within the trees.

What you are suggesting is indeed so complex it is clearly out of scope for this PR.
I like your idea of sending only partial updates to the client rather than the entire tree every time, but that does not need to be the first time we will expose the verification tree. It's merely an low-level optimization that we need to carefully consider. This optimization would first consist in creating and exposing a new API to send partial verification diagnostics update. If the client does not define this API, then the server will just send the whole tree every time instead of updates.

Hence, for now and in order to display related ranges, we can keep it simple and just focus on the definition of the tree itself, and keep sending the verification tree to the client in the scope of this PR.
There is always an option to deactivate this altogether if one suffers from performance, but so far I haven't seen performance issue even on large files. The biggest problem currently is merely from the language server not canceling requests fast enough when it's stuck, which we should address separatedly.